src/imap/imap-fetch.c

```diff
@@ -916,6 +916,66 @@ static bool imap_fetch_savedate_init(struct imap_fetch_init_context *ctx)
 	return TRUE;
 }
 
+static int fetch_emailid(struct imap_fetch_context *ctx, struct mail *mail,
+		      void *context ATTR_UNUSED)
+{
+	const char *value;
+
+	if (mail_get_special(mail, MAIL_FETCH_GUID, &value) < 0)
+		return -1;
+
+	str_printfa(ctx->state.cur_str, "EMAILID (%s) ", value);
+	return 1;
+}
+
+static bool fetch_emailid_init(struct imap_fetch_init_context *ctx)
+{
+	ctx->fetch_ctx->fetch_data |= MAIL_FETCH_GUID;
+	imap_fetch_add_handler(ctx, IMAP_FETCH_HANDLER_FLAG_BUFFERED,
+			       "", fetch_emailid, NULL);
+	return TRUE;
+}
+
+static int fetch_threadid(struct imap_fetch_context *ctx, struct mail *mail,
+		      void *context ATTR_UNUSED)
+{
+	const char *value;
+
+	if (mail_get_special(mail, MAIL_FETCH_GUID, &value) < 0)
+		return -1;
+
+	str_printfa(ctx->state.cur_str, "THREADID (%s) ", value);
+	return 1;
+}
+
+static bool fetch_threadid_init(struct imap_fetch_init_context *ctx)
+{
+	ctx->fetch_ctx->fetch_data |= MAIL_FETCH_GUID;
+	imap_fetch_add_handler(ctx, IMAP_FETCH_HANDLER_FLAG_BUFFERED,
+			       "", fetch_threadid, NULL);
+	return TRUE;
+}
+
+static int fetch_mailboxid(struct imap_fetch_context *ctx, struct mail *mail,
+			   void *context ATTR_UNUSED)
+{
+	struct mailbox_metadata metadata;
+
+	if (mailbox_get_metadata(mail->box, MAILBOX_METADATA_GUID, &metadata) < 0)
+		return -1;
+
+	str_printfa(ctx->state.cur_str, "MAILBOXID (%s) ",
+			guid_128_to_string(metadata.guid));
```

Comment:
There's no MAILBOXID for FETCH, only STATUS.

Reply:
yes, but there is already X-MAILBOX, this could be more standardized name and usable for clients when using virtual \All mailbox